From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: George T Kramer <george.t.kramer@intel.com>
Date: Fri, 31 Mar 2017 14:42:59 -0700
Subject: [PATCH] Remove RC4 support for OpenSSL

---
 common/ssl_calls.c | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/common/ssl_calls.c b/common/ssl_calls.c
index d1003b8..8be10d5 100644
--- a/common/ssl_calls.c
+++ b/common/ssl_calls.c
@@ -26,7 +26,6 @@
 #include <stdlib.h> /* needed for openssl headers */
 #include <openssl/ssl.h>
 #include <openssl/err.h>
-#include <openssl/rc4.h>
 #include <openssl/md5.h>
 #include <openssl/sha.h>
 #include <openssl/hmac.h>
@@ -127,7 +126,7 @@ ssl_finish(void)
 void *
 ssl_rc4_info_create(void)
 {
-    return g_malloc(sizeof(RC4_KEY), 1);
+    return NULL;
 }
 
 /*****************************************************************************/
@@ -141,14 +140,12 @@ ssl_rc4_info_delete(void *rc4_info)
 void
 ssl_rc4_set_key(void *rc4_info, char *key, int len)
 {
-    RC4_set_key((RC4_KEY *)rc4_info, len, (tui8 *)key);
 }
 
 /*****************************************************************************/
 void
 ssl_rc4_crypt(void *rc4_info, char *data, int len)
 {
-    RC4((RC4_KEY *)rc4_info, len, (tui8 *)data, (tui8 *)data);
 }
 
 /* sha1 stuff */
