diff -Naur xrdp-0.9.9-orig/common/os_calls.c xrdp-0.9.9/common/os_calls.c
--- xrdp-0.9.9-orig/common/os_calls.c	2018-12-20 02:15:29.000000000 +0000
+++ xrdp-0.9.9/common/os_calls.c	2019-02-01 20:30:56.434569041 +0000
@@ -3823,3 +3823,32 @@
     return 0;
 }
 
+void g_snprintf_config(char *dest, int size, const char *fmt, ...)
+{
+    char *user_fmtd = alloca(size);
+    va_list ap;
+    int err;
+
+    va_start(ap, fmt);
+    err = vsnprintf(user_fmtd, size, fmt, ap);
+    va_end(ap);
+    if (err >= size)
+        goto out;
+
+    if (snprintf(dest, size, XRDP_CFG_PATH"/%s", user_fmtd) >= size)
+        goto out;
+    if (!access(dest, R_OK))
+        return;
+
+    if (snprintf(dest, size, "/usr/share/defaults/xrdp/%s", user_fmtd) >= size)
+        goto out;
+    if (!access(dest, R_OK))
+        return;
+
+    snprintf(dest, size, XRDP_CFG_PATH"/%s", user_fmtd);
+    return;
+
+out:
+    fprintf(stderr, "Buffer exceeds given size, aborting");
+    abort();
+}
\ No newline at end of file
diff -Naur xrdp-0.9.9-orig/common/os_calls.h xrdp-0.9.9/common/os_calls.h
--- xrdp-0.9.9-orig/common/os_calls.h	2018-12-20 02:15:29.000000000 +0000
+++ xrdp-0.9.9/common/os_calls.h	2019-02-01 20:30:56.434569041 +0000
@@ -183,6 +183,7 @@
 int      g_shmdt(const void *shmaddr);
 int      g_gethostname(char *name, int len);
 int      g_mirror_memcpy(void *dst, const void *src, int len);
+void	 g_snprintf_config(char *dest, int size, const char *fmt, ...) printflike(3, 4);
 
 /* glib-style wrappers */
 #define g_new(struct_type, n_structs) \
diff -Naur xrdp-0.9.9-orig/libxrdp/xrdp_rdp.c xrdp-0.9.9/libxrdp/xrdp_rdp.c
--- xrdp-0.9.9-orig/libxrdp/xrdp_rdp.c	2018-12-25 04:45:45.000000000 +0000
+++ xrdp-0.9.9/libxrdp/xrdp_rdp.c	2019-02-01 20:30:52.456523206 +0000
@@ -60,7 +60,7 @@
     items->auto_free = 1;
     values = list_create();
     values->auto_free = 1;
-    g_snprintf(cfg_file, 255, "%s/xrdp.ini", XRDP_CFG_PATH);
+    g_snprintf_config(cfg_file, 255, "xrdp.ini");
     DEBUG(("cfg_file %s", cfg_file));
     file_by_name_read_section(cfg_file, "globals", items, values);
 
diff -Naur xrdp-0.9.9-orig/libxrdp/xrdp_sec.c xrdp-0.9.9/libxrdp/xrdp_sec.c
--- xrdp-0.9.9-orig/libxrdp/xrdp_sec.c	2018-12-25 04:45:45.000000000 +0000
+++ xrdp-0.9.9/libxrdp/xrdp_sec.c	2019-02-01 20:30:52.456523206 +0000
@@ -260,7 +260,7 @@
         client_info->keyboard_subtype = 1;
     }
 
-    g_snprintf(keyboard_cfg_file, 255, "%s/xrdp_keyboard.ini", XRDP_CFG_PATH);
+    g_snprintf_config(keyboard_cfg_file, 255, "xrdp_keyboard.ini");
     LLOGLN(10, ("keyboard_cfg_file %s", keyboard_cfg_file));
 
     fd = g_file_open(keyboard_cfg_file);
@@ -2302,7 +2302,7 @@
             items->auto_free = 1;
             values = list_create();
             values->auto_free = 1;
-            g_snprintf(key_file, 255, "%s/rsakeys.ini", XRDP_CFG_PATH);
+            g_snprintf_config(key_file, 255, "rsakeys.ini");
 
             if (file_by_name_read_section(key_file, "keys", items, values) != 0)
             {
diff -Naur xrdp-0.9.9-orig/sesman/chansrv/chansrv.c xrdp-0.9.9/sesman/chansrv/chansrv.c
--- xrdp-0.9.9-orig/sesman/chansrv/chansrv.c	2018-12-25 04:45:45.000000000 +0000
+++ xrdp-0.9.9/sesman/chansrv/chansrv.c	2019-02-01 20:30:52.456523206 +0000
@@ -1636,7 +1636,7 @@
     values = list_create();
     values->auto_free = 1;
     g_use_unix_socket = 0;
-    g_snprintf(filename, 255, "%s/sesman.ini", XRDP_CFG_PATH);
+    g_snprintf_config(filename, 255, "sesman.ini");
 
     if (file_by_name_read_section(filename, "Globals", names, values) == 0)
     {
diff -Naur xrdp-0.9.9-orig/sesman/chansrv/chansrv_fuse.c xrdp-0.9.9/sesman/chansrv/chansrv_fuse.c
--- xrdp-0.9.9-orig/sesman/chansrv/chansrv_fuse.c	2018-12-25 04:45:45.000000000 +0000
+++ xrdp-0.9.9/sesman/chansrv/chansrv_fuse.c	2019-02-01 20:30:52.456523206 +0000
@@ -379,7 +379,7 @@
     items->auto_free = 1;
     values = list_create();
     values->auto_free = 1;
-    g_snprintf(cfg_file, 255, "%s/sesman.ini", XRDP_CFG_PATH);
+    g_snprintf_config(cfg_file, 255, "sesman.ini");
     file_by_name_read_section(cfg_file, "Chansrv", items, values);
     for (index = 0; index < items->count; index++)
     {
diff -Naur xrdp-0.9.9-orig/sesman/config.c xrdp-0.9.9/sesman/config.c
--- xrdp-0.9.9-orig/sesman/config.c	2018-10-25 06:20:43.000000000 +0000
+++ xrdp-0.9.9/sesman/config.c	2019-02-01 20:30:52.456523206 +0000
@@ -48,7 +48,7 @@
     struct list *param_v;
     char cfg_file[256];
 
-    g_snprintf(cfg_file, 255, "%s/sesman.ini", XRDP_CFG_PATH);
+    g_snprintf_config(cfg_file, 255, "sesman.ini");
     fd = g_file_open(cfg_file);
 
     if (-1 == fd)
diff -Naur xrdp-0.9.9-orig/sesman/sesman.c xrdp-0.9.9/sesman/sesman.c
--- xrdp-0.9.9-orig/sesman/sesman.c	2018-10-25 06:20:43.000000000 +0000
+++ xrdp-0.9.9/sesman/sesman.c	2019-02-01 20:30:52.456523206 +0000
@@ -317,7 +317,7 @@
         config_dump(g_cfg);
     }
 
-    g_snprintf(cfg_file, 255, "%s/sesman.ini", XRDP_CFG_PATH);
+    g_snprintf_config(cfg_file, 255, "sesman.ini");
 
     /* starting logging subsystem */
     log_error = log_start(cfg_file, "xrdp-sesman");
diff -Naur xrdp-0.9.9-orig/sesman/sig.c xrdp-0.9.9/sesman/sig.c
--- xrdp-0.9.9-orig/sesman/sig.c	2018-10-19 07:09:19.000000000 +0000
+++ xrdp-0.9.9/sesman/sig.c	2019-02-01 20:30:52.457523218 +0000
@@ -101,7 +101,7 @@
     /* replace old config with newly read one */
     g_cfg = cfg;
 
-    g_snprintf(cfg_file, 255, "%s/sesman.ini", XRDP_CFG_PATH);
+    g_snprintf_config(cfg_file, 255, "sesman.ini");
 
     /* start again logging subsystem */
     error = log_start(cfg_file, "xrdp-sesman");
diff -Naur xrdp-0.9.9-orig/xrdp/lang.c xrdp-0.9.9/xrdp/lang.c
--- xrdp-0.9.9-orig/xrdp/lang.c	2018-10-19 07:09:19.000000000 +0000
+++ xrdp-0.9.9/xrdp/lang.c	2019-02-01 20:30:52.457523218 +0000
@@ -237,21 +237,21 @@
     filename = (char *)g_malloc(256, 0);
 
     /* check if there is a keymap file e.g. km-e00100411.ini */
-    g_snprintf(filename, 255, "%s/km-%08x.ini", XRDP_CFG_PATH, keylayout);
+    g_snprintf_config(filename, 255, "km-%08x.ini", keylayout);
 
     /* if the file does not exist, use only lower 16 bits instead */
     if (!g_file_exist(filename))
     {
         log_message(LOG_LEVEL_INFO, "Cannot find keymap file %s", filename);
         /* e.g. km-00000411.ini */
-        g_snprintf(filename, 255, "%s/km-%08x.ini", XRDP_CFG_PATH, basic_key_layout);
+        g_snprintf_config(filename, 255, "km-%08x.ini", basic_key_layout);
     }
 
     /* finally, use 'en-us' */
     if (!g_file_exist(filename))
     {
         log_message(LOG_LEVEL_INFO, "Cannot find keymap file %s", filename);
-        g_snprintf(filename, 255, "%s/km-00000409.ini", XRDP_CFG_PATH);
+        g_snprintf_config(filename, 255, "km-00000409.ini");
     }
 
     if (g_file_exist(filename))
diff -Naur xrdp-0.9.9-orig/xrdp/xrdp.c xrdp-0.9.9/xrdp/xrdp.c
--- xrdp-0.9.9-orig/xrdp/xrdp.c	2018-10-25 06:20:43.000000000 +0000
+++ xrdp-0.9.9/xrdp/xrdp.c	2019-02-01 20:30:52.457523218 +0000
@@ -394,7 +394,7 @@
         return 1;
     }
 
-    g_snprintf(key_file, 255, "%s/rsakeys.ini", XRDP_CFG_PATH);
+    g_snprintf_config(key_file, 255, "rsakeys.ini");
     if (!g_file_exist(key_file))
     {
         g_writeln("File %s is missing, create it using xrdp-keygen", key_file);
@@ -428,7 +428,7 @@
         DEBUG(("Argument %i - %s", test, argv[test]));
     }
 
-    g_snprintf(cfg_file, 255, "%s/xrdp.ini", XRDP_CFG_PATH);
+    g_snprintf_config(cfg_file, 255, "xrdp.ini");
 
     startup_params = (struct xrdp_startup_params *)
                      g_malloc(sizeof(struct xrdp_startup_params), 1);
diff -Naur xrdp-0.9.9-orig/xrdp/xrdp_listen.c xrdp-0.9.9/xrdp/xrdp_listen.c
--- xrdp-0.9.9-orig/xrdp/xrdp_listen.c	2018-10-25 06:20:43.000000000 +0000
+++ xrdp-0.9.9/xrdp/xrdp_listen.c	2019-02-01 20:30:52.457523218 +0000
@@ -169,7 +169,7 @@
     /* Default to all */
     g_strncpy(address, "0.0.0.0", address_bytes - 1);
     /* see if port or address is in xrdp.ini file */
-    g_snprintf(cfg_file, 255, "%s/xrdp.ini", XRDP_CFG_PATH);
+    g_snprintf_config(cfg_file, 255, "xrdp.ini");
     fd = g_file_open(cfg_file);
     *mode = TRANS_MODE_TCP;
     *tcp_nodelay = 0 ;
diff -Naur xrdp-0.9.9-orig/xrdp/xrdp_login_wnd.c xrdp-0.9.9/xrdp/xrdp_login_wnd.c
--- xrdp-0.9.9-orig/xrdp/xrdp_login_wnd.c	2018-10-25 07:00:14.000000000 +0000
+++ xrdp-0.9.9/xrdp/xrdp_login_wnd.c	2019-02-01 20:30:52.457523218 +0000
@@ -568,7 +568,7 @@
     section_names->auto_free = 1;
     section_values = list_create();
     section_values->auto_free = 1;
-    g_snprintf(cfg_file, 255, "%s/xrdp.ini", XRDP_CFG_PATH);
+    g_snprintf_config(cfg_file, 255, "xrdp.ini");
     fd = g_file_open(cfg_file); /* xrdp.ini */
 
     if (fd < 0)
@@ -873,7 +873,7 @@
     globals->ls_btn_cancel_height = 30;
 
     /* open xrdp.ini file */
-    g_snprintf(buf, 255, "%s/xrdp.ini", XRDP_CFG_PATH);
+    g_snprintf_config(buf, 255, "xrdp.ini");
     if ((fd = g_file_open(buf)) < 0)
     {
         log_message(LOG_LEVEL_ERROR,"load_config: Could not read "
diff -Naur xrdp-0.9.9-orig/xrdp/xrdp_mm.c xrdp-0.9.9/xrdp/xrdp_mm.c
--- xrdp-0.9.9-orig/xrdp/xrdp_mm.c	2018-12-25 04:45:45.000000000 +0000
+++ xrdp-0.9.9/xrdp/xrdp_mm.c	2019-02-01 20:30:52.457523218 +0000
@@ -1602,7 +1602,7 @@
     /* default to port 3350 */
     g_strncpy(port, "3350", port_bytes - 1);
     /* see if port is in sesman.ini file */
-    g_snprintf(cfg_file, 255, "%s/sesman.ini", XRDP_CFG_PATH);
+    g_snprintf_config(cfg_file, 255, "sesman.ini");
     fd = g_file_open(cfg_file);
 
     if (fd >= 0)
diff -Naur xrdp-0.9.9-orig/xrdp/xrdp_wm.c xrdp-0.9.9/xrdp/xrdp_wm.c
--- xrdp-0.9.9-orig/xrdp/xrdp_wm.c	2018-12-25 04:45:45.000000000 +0000
+++ xrdp-0.9.9/xrdp/xrdp_wm.c	2019-02-01 20:30:52.458523229 +0000
@@ -414,7 +414,7 @@
     self->background = HCOLOR(self->screen->bpp, 0x000000);
 
     /* now load them from the globals in xrdp.ini if defined */
-    g_snprintf(cfg_file, 255, "%s/xrdp.ini", XRDP_CFG_PATH);
+    g_snprintf_config(cfg_file, 255, "xrdp.ini");
     fd = g_file_open(cfg_file);
 
     if (fd >= 0)
@@ -580,7 +580,7 @@
     names->auto_free = 1;
     values = list_create();
     values->auto_free = 1;
-    g_snprintf(cfg_file, 255, "%s/xrdp.ini", XRDP_CFG_PATH);
+    g_snprintf_config(cfg_file, 255, "xrdp.ini");
     if (file_by_name_read_section(cfg_file, "Channels", names, values) == 0)
     {
         int error;
@@ -648,7 +648,7 @@
          * NOTE: this should eventually be accessed from self->xrdp_config
          */
 
-        g_snprintf(cfg_file, 255, "%s/xrdp.ini", XRDP_CFG_PATH);
+        g_snprintf_config(cfg_file, 255, "xrdp.ini");
         fd = g_file_open(cfg_file); /* xrdp.ini */
         if (fd != -1)
         {
